content/posts/2022-05-13-calculating-type-sets.markdown

```diff
@@ -0,0 +1,607 @@
+---
+date: "2022-05-12 08:00:00"
+tags:
+- golang
+- programming
+title: "Calculating type sets is harder than you think"
+summary: I try to explain why the problem of calculating the exact contents of
+  a type set in Go is harder than it might seem on the surface.
+---
+
+<script defer crossorigin="anonymous" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-MML-AM_CHTML"></script>
+
+Go 1.18 added the biggest and probably one of the most requested features of
+all time to the language: [Generics](https://go.dev/doc/go1.18#generics). If
+you want a comprehensive introduction to the topic, there are many out there
+and I would personally [recommend this talk I gave at the Frankfurt Gopher
+Meetup](https://www.youtube.com/watch?v=QP6v-Q5Foek).
+
+This blog post is not an introduction to generics, though. It is about [this
+sentence from the spec](https://go.dev/ref/spec#Operands):
+
+> Implementation restriction: A compiler need not report an error if an
+> operand's type is a type parameter with an empty type set.
+
+This restriction might seem strange to you. After all, if a type set is empty,
+it seems very helpful to report that to the user, because they obviously made a
+mistake - an empty type set can never be used as a constraint, as the function
+using it could never be instantiated.
+
+I want to explain the reason that sentence is there and also go into a couple
+of related design decisions of the generics design. I'm trying to be expansive
+in my explanation, which means that you should not *need* any special knowledge
+to understand it. It also means, some of the information might be boring to
+you - feel free to skip the corresponding sections.
+
+The reason that sentence is in the Go spec, is that it turns out to be hard to
```

Comment:
If you did follow the "question" and "answer" proposal above, this paragraph, minus the last sentence, goes after the answer, before the paragraph "This article..."